onnxruntime/core/optimizer/graph_transformer_utils.cc

```diff
@@ -324,6 +316,19 @@ InlinedVector<std::unique_ptr<GraphTransformer>> GenerateTransformers(
       // while we can fuse more activation.
       transformers.emplace_back(std::make_unique<ConvAddActivationFusion>(cpu_ep));
 #endif
+
+#ifdef ENABLE_TRAINING
+      // Put memory optimization transformer in level 3 (which is done after most of fusions are done) by intention.
```

Comment:
Level 3 optimizers are meant for layout optimizations... We should not add this one to level 3. We need to find another way to prevent this issue.

Reply:
Few alternatives so far:
1. Leave it as it is. New fusion after the memory optimizer will potentially dis-order the expected recompute subgraphs. And this is because we are using node priority to control the execution order. If we want to solve this problem elegantly, we need do some work on the topo sort to make those recomputed node work as expected. A possible solution is, we leverage Function, say we put all detected subgraph nodes into a Recompute node which has a function during the memory optimzier transformation. Not sure whether it is feasible. 
2. Add special logic in TransformGraph in inference_session.cc, just as we did for insert_cast, insert_copy transformers, we put ours there two. This makes sense because you know insert_cast/insert_copy could also dis-order the execution of recomputed nodes.

Reply:
Option 2 or putting into layer 3 is simplest way as a short term. And compare those two, putting into layer 3 restricted with enable_traing might be less concerned. 

Do you have other suggestions?

Reply:
As suggested offline, I move it back to Level 2 and add a comment/todo, I will find some way to refine.